app/namespaces/test_adjustment/test_adjustment_service.py

```diff
@@ -0,0 +1,266 @@
+# monte-carlo version of RG estimator,
+# accounting for uncertainty in se and sp estimates
+import pickle
+from math import log
+from hashlib import md5
+from typing import Dict, Tuple, Union
+
+import pandas as pd
+import pystan
+import arviz
+from statsmodels.stats.proportion import proportion_confint
+from marshmallow import Schema, fields, ValidationError
+
+from app.utils.test_adjustment import bastos_estimates, testadj_model_code
+
+
+def logit(p, tol=1e-3):
+    # logit function; well defined for p in open interval (0,1)
+
+    # constrain the probability to the range [tol, 1-tol]
+    # to avoid NaNs if p = 0 or p = 1
+    p = max(p, tol)
+    p = min(p, 1 - tol)
+
+    # return the logit of the constrained probability
+    return log(p / (1 - p))
+
+
+def result_is_bounded(median_adj_prev, raw_prev):
+    # check whether the adjusted result is close to the raw value
+    # e.g., log odds within 1
+    logit_delta = logit(median_adj_prev) - logit(raw_prev)
+    adjusted_closeto_raw = abs(logit_delta) < 1
+
+    # permit the adjusted result if adjusted and raw are both < 0.5
+    # or if both are > 0.5
+    both_below_maxsmall = (median_adj_prev <= 0.5) and (raw_prev <= 0.5)
+    both_above_minbig = (median_adj_prev >= 0.5) and (raw_prev >= 0.5)
+
+    return (adjusted_closeto_raw or both_below_maxsmall or both_above_minbig)
+
+
+def validate_against_schema(input_payload, schema):
+    try:
+        payload = schema.load(input_payload)
+    except ValidationError as err:
+        return err.messages
+    return payload
+
+
+class ModelParamsSchema(Schema):
+    n_prev_obs = fields.Integer()
+    y_prev_obs = fields.Float()
+    n_se = fields.Integer()
+    y_se = fields.Float()
+    n_sp = fields.Integer()
+    y_sp = fields.Float()
+
+
+class TestAdjHandler:
+    def __init__(self, model_code=testadj_model_code, model_name='testadj_binomial_se_sp',
+                 execution_params={}):
+        self.TESTADJ_MODEL = self.get_stan_model_cache(model_code=model_code, model_name=model_name)
+        self.TESTADJ_MODEL_NAME = model_name
+        # Parse execution params
+        self.n_iter = execution_params.get('n_iter', 2000)
```

Comment:
Might be able to drop this to 1e3 if this is still running slow. Let's leave it as is for now